ca/certificate-authority.go

```diff
@@ -215,176 +217,397 @@ func (ca *CertificateAuthorityImpl) RevokeCertificate(serial string, reasonCode
+
+	if ca.NotAfter.Before(notAfter) {
+		err = core.InternalServerError("Cannot issue a certificate that expires after the intermediate certificate.")
+		// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
+		ca.Log.AuditErr(err)
+		return err
+	}
+
+	if req.Expires.Before(ca.Clk.Now()) {
+		err = core.InternalServerError("Cannot issue on an expired request.")
+		// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
+		ca.Log.AuditErr(err)
+		return err
+	}
+
+	err = ca.SA.UpdateCertificateRequestStatus(requestID, core.StatusPending)
+	if err != nil {
+		return err
+	}
+
+	signRequest := signer.SignRequest{
+		Request: csrPEM,
+		Profile: ca.profile,
+		Hosts:   hostNames,
+		Subject: &signer.Subject{
+			CN: commonName,
+		},
+	}
+
+	// TODO(rlb): Run this in a goroutine to make it async
+	ca.signCertificate(csr, signRequest, requestID, logEventID)
+
+	logEventResult = "successful"
+	return nil
+}
+
+func (ca *CertificateAuthorityImpl) signCertificate(csr *x509.CertificateRequest, signRequest signer.SignRequest, requestID string, logEventID string) {
+	logEventResult := "error"
+	logEvent := blog.CertificateRequestEvent{
+		ID:          logEventID + ".signCertificate",
+		RequestTime: time.Now(),
+	}
+
+	// No matter what, log the request
+	defer func() {
+		// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
+		ca.Log.AuditObject(fmt.Sprintf("Certificate request - %s", logEventResult), logEvent)
+	}()
+
+	// Generate a new random serial number
 	// We want 136 bits of random number, plus an 8-bit instance id prefix.
 	const randBits = 136
 	serialBytes := make([]byte, randBits/8+1)
 	serialBytes[0] = byte(ca.Prefix)
-	_, err = rand.Read(serialBytes[1:])
+	_, err := rand.Read(serialBytes[1:])
 	if err != nil {
 		err = core.InternalServerError(err.Error())
 		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
-		ca.log.Audit(fmt.Sprintf("Serial randomness failed, err=[%v]", err))
-		return emptyCert, err
+		ca.Log.Audit(fmt.Sprintf("Serial randomness failed, err=[%v]", err))
+		return
 	}
 	serialHex := hex.EncodeToString(serialBytes)
 	serialBigInt := big.NewInt(0)
 	serialBigInt = serialBigInt.SetBytes(serialBytes)
+	signRequest.Serial = serialBigInt
 
-	// Send the cert off for signing
-	req := signer.SignRequest{
-		Request: csrPEM,
-		Profile: ca.profile,
-		Hosts:   hostNames,
-		Subject: &signer.Subject{
-			CN: commonName,
-		},
-		Serial: serialBigInt,
-	}
-
-	certPEM, err := ca.Signer.Sign(req)
+	certPEM, err := ca.Signer.Sign(signRequest)
 	if err != nil {
 		err = core.InternalServerError(err.Error())
 		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
-		ca.log.Audit(fmt.Sprintf("Signer failed, rolling back: serial=[%s] err=[%v]", serialHex, err))
-		return emptyCert, err
+		ca.Log.Audit(fmt.Sprintf("Signer failed, rolling back: serial=[%s] err=[%v]", serialHex, err))
+		ca.SA.UpdateCertificateRequestStatus(requestID, core.StatusInvalid)
```

Comment:
We should have the ability to attach a problem status to a cert request, so that it can be returned in an error document.  But maybe that's a separate PR.